src/main/java/teammates/common/util/TimeKeeper.java

```diff
@@ -0,0 +1,29 @@
+package teammates.common.util;
+
+import java.util.Date;
+
+import com.google.apphosting.api.ApiProxy;
+import com.google.apphosting.api.DeadlineExceededException;
+
+public class TimeKeeper {
+    private static final long REMAINING_TIME_THRESHOLD = 5000;
+    private static final long TIME_BETWEEN_CHECKS = 5000;
+    private Date lastCalledDateForEnoughTimeMethod;
+    
+    public void hasEnoughTimeThrowException() throws DeadlineExceededException {
+        Date now = new Date();
+        if (lastCalledDateForEnoughTimeMethod != null
+                && now.getTime() - lastCalledDateForEnoughTimeMethod.getTime() <= TIME_BETWEEN_CHECKS) {
+            return;
+        }
+        lastCalledDateForEnoughTimeMethod = new Date();
+        boolean hasEnoughTime = hasEnoughTime();
+        if (!hasEnoughTime) {
+            throw new DeadlineExceededException();
+        }
```

Comment:
Why not calculate the deadline expiry time (i.e. current time + remaining time) and store it internally? After that, no need to call the API at all.